rpc/src/rpc.rs

```diff
@@ -552,51 +558,130 @@ impl JsonRpcRequestProcessor {
+                            epoch, partition_data_address
+                        ),
+                        data: None,
+                    })?;
+            let EpochRewardsPartitionDataVersion::V0(partition_data) =
+                bincode::deserialize(partition_data_account.data())
+                    .map_err(|_| Error::internal_error())?;
+            let hasher = EpochRewardsHasher::new(
+                partition_data.num_partitions,
+                &partition_data.parent_blockhash,
+            );
+            let mut partition_index_addresses: HashMap<usize, Vec<String>> = HashMap::new();
+            for address in addresses.iter() {
+                let address_string = address.to_string();
+                // Skip this address if (Voting) rewards were already found in
+                // the first block of the epoch
+                if !reward_map.contains_key(&address_string) {
+                    let partition_index = hasher.clone().hash_address_to_partition(address);
+                    partition_index_addresses
+                        .entry(partition_index)
+                        .and_modify(|list| list.push(address_string.clone()))
+                        .or_insert(vec![address_string]);
+                }
+            }
 
-        let reward_hash: HashMap<String, Reward> = first_confirmed_block
-            .rewards
-            .unwrap_or_default()
-            .into_iter()
-            .filter_map(|reward| match reward.reward_type? {
-                RewardType::Staking | RewardType::Voting => addresses
-                    .contains(&reward.pubkey)
-                    .then(|| (reward.clone().pubkey, reward)),
-                _ => None,
-            })
-            .collect();
+            let block_list = self
+                .get_blocks_with_limit(
+                    first_slot_in_epoch,
+                    partition_data.num_partitions + 1,
+                    config.commitment,
+                )
+                .await?;
+
+            for (partition_index, addresses) in partition_index_addresses.iter() {
+                let slot = *block_list
+                    .get(partition_index.saturating_add(1))
+                    .ok_or_else(Error::internal_error)?;
+                let Ok(Some(block)) = self
+                    .get_block(
+                        slot,
+                        Some(RpcBlockConfig::rewards_with_commitment(config.commitment).into()),
+                    )
+                    .await
+                else {
+                    return Err(RpcCustomError::BlockNotAvailable { slot }.into());
+                };
+                let index_reward_map: HashMap<String, (Reward, Slot)> = block
+                    .rewards
+                    .unwrap_or_default()
+                    .into_iter()
+                    .filter_map(|reward| match reward.reward_type? {
+                        RewardType::Staking | RewardType::Voting => addresses
```

Comment:
This one can be changed, yes.

Reply:
Given the dereplicode refactoring, I'm not sure this is worth doing. It's a little prickly to match only one or the other type after feature activation, but both before.

Reply:
> It's a little prickly to match only one or the other type after feature activation, but both before.

mmm... is it tho? split the `filter_map()` to a discrete `filter()` and `map()`, pass in different filter closures, works no?

Reply:
Fine, I've pushed that. I don't think it gains us much, though, since after feature activation there won't be any RewardType::Staking rewards in the first block to check.